### PR TITLE
IN 720 config module

### PR DESCRIPTION
**Pull request by ehanson8**
### What does this PR do?
Creates functions to manage configuration values.

### Helpful background context
With the simplified configuration requirements, I chose to use 2 functions rather than a class. I also assumed that every variable must be present (even if the value is `None`) given that any missing variables would indicate an infrastructure problem. I also reworked the `configure_sentry` function so there are no duplicate calls to ENV variables. Happy to hear any feedback on these choices.

### How can a reviewer manually see the effects of these changes?
Set all required ENV variables in an `.env` file, see `README.md` for the list. Before creating the last variable, run the program with `pipenv run patronload` to see the error for a missing variable. Add the final variable and run the program again to see a completed run.

### Includes new or updated dependencies?
NO

### What are the relevant tickets?
* https://mitlibraries.atlassian.net/browse/IN-720

### Developer

- [x] All new ENV is documented in README (or there is none)
- [x] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer

- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: patronload/config.py [hunk 1]
```diff
@@ -1,8 +1,22 @@
 import logging
 import os
+from typing import Optional
 
 import sentry_sdk
 
+EXPECTED_CONFIG_VARIABLES = [
```
Comment (hakbailey) [inline patronload/config.py:7]: What I would probably do instead here is set default config values and then in the `load_config_values()` function update them from ENV. If a value isn't present in ENV then this will just use the default. We don't actually need all these in every environment, so setting a default value of `None` for those is fine.

File: patronload/config.py [hunk 2]
```diff
@@ -24,10 +38,32 @@ def configure_logger(logger: logging.Logger, verbose: bool) -> str:
     )
 
 
-def configure_sentry() -> str:
-    env = os.getenv("WORKSPACE")
-    sentry_dsn = os.getenv("SENTRY_DSN")
+def configure_sentry(env: str, sentry_dsn: str) -> str:
```
Comment (hakbailey) [inline patronload/config.py:41]: I think the `sentry_dsn` type should be `Optional[str]`?

File: tests/test_config.py [hunk 3]
```diff
@@ -17,19 +24,49 @@ def test_configure_logger_verbose():
     assert result == "Logger 'tests.test_config' configured with level=DEBUG"
 
 
-def test_configure_sentry_no_env_variable(monkeypatch):
-    monkeypatch.delenv("SENTRY_DSN", raising=False)
-    result = configure_sentry()
+def test_configure_sentry_env_variable_is_none():
```
Comment (hakbailey) [inline tests/test_config.py:27]: We should also include a test for the `sentry_dsn` value being the string `"None"`, since that's what it will be when read from the Dev1 env.

File: README.md [hunk 4]
```diff
@@ -17,8 +17,8 @@ Creates Alma-compliant XML patron profiles from information extracted from the D
 - `DATA_WAREHOUSE_HOST` = The host for the Data Warehouse database.
 - `DATA_WAREHOUSE_PORT` = The port for the Data Warehouse database.
 - `DATA_WAREHOUSE_SID` = The system identifier for the Data Warehouse database instance.
+- `LOG_LEVEL` = The log level for the `alma-patronload` application
 - `S3_BUCKET_NAME` = The S3 bucket in which files are deposited.
 - `S3_PATH` = The path to the S3 bucket in which files are deposited.
```
Comment (hakbailey) [inline README.md:22]: This is really the file path prefix for files deposited to the S3 bucket, right?

File: patronload/config.py [hunk 5]
```diff
@@ -24,10 +38,32 @@ def configure_logger(logger: logging.Logger, verbose: bool) -> str:
...
+            config_values[expected_config_variable] = config_value
+    return config_values
+
+
+def get_required_env_variable(variable_name: str) -> Optional[str]:
+    try:
+        return os.environ[variable_name]
+    except KeyError as error:
+        raise KeyError(
+            f"Missing variable: '{variable_name}' is required for all environments"
```
Comment (hakbailey) [inline patronload/config.py:68]: We explicitly state in the README that not all of the variables are required in all envs, so we either need to update that or change the requirement here. I would suggest the latter, as they really aren't all required in all envs.

**Review by hakbailey — COMMENTED** [review]
See inline suggestions and questions.

File: patronload/cli.py [hunk 6]
```diff
@@ -17,7 +17,15 @@ def main(verbose: bool) -> None:
     start_time = perf_counter()
     root_logger = logging.getLogger()
     logger.info(configure_logger(root_logger, verbose))
-    logger.info(configure_sentry())
+    config_values = load_config_values()
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply patronload/cli.py:20]: Yes, that was the plan, are there issues with that approach?

File: patronload/config.py [hunk 7]
```diff
@@ -31,3 +35,31 @@ def configure_sentry() -> str:
         sentry_sdk.init(sentry_dsn, environment=env)
         return f"Sentry DSN found, exceptions will be sent to Sentry with env={env}"
     return "No Sentry DSN found, exceptions will not be sent to Sentry"
+
+
+def load_config_values() -> dict:
```
Comment (adamshire123) [inline patronload/config.py:40]: It looks like the default config values never get used, since get_required_env_variable raises an error if it can't find the value.

Comment (ehanson8) [reply patronload/config.py:40]: You can use something like the `test_env` fixture in `conftest.py` in my latest PR,  I thought I didn't need after adding the default values but it turns out I did

File: patronload/cli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hakbailey) [reply patronload/cli.py:15]: Not necessarily but depending on how the app is structured it could get cumbersome. Fine to leave it as is and change later if that happens.

File: patronload/config.py (hunk 7 above)
Comment (hakbailey) [reply patronload/config.py:40]: Do we actually need the check for required env variables in this app? I'm not sure we do. It was present in alma-scripts because 1) the app and infrastructure were complicated, and 2) the sap invoice process has some very strict configuration requirements. The patronload process is more straightforward and simply won't work correctly if the required variables aren't present.

File: README.md [hunk 8]
```diff
@@ -17,8 +17,10 @@ Creates Alma-compliant XML patron profiles from information extracted from the D
 - `DATA_WAREHOUSE_HOST` = The host for the Data Warehouse database.
 - `DATA_WAREHOUSE_PORT` = The port for the Data Warehouse database.
 - `DATA_WAREHOUSE_SID` = The system identifier for the Data Warehouse database instance.
-- `S3_BUCKET_NAME` = The S3 bucket in which files are deposited.
-- `S3_PATH` = The path to the S3 bucket in which files are deposited.
 - `LOG_LEVEL` = The log level for the `alma-patronload` application
```
Comment (hakbailey) [inline README.md:20]: `LOG_LEVEL` should also be optional since it defaults to INFO.

File: patronload/cli.py [hunk 9]
```diff
@@ -4,20 +4,24 @@
...
-@click.option(
-    "-v", "--verbose", is_flag=True, help="Pass to log at debug level instead of info"
-)
-def main(verbose: bool) -> None:
+def main() -> None:
     start_time = perf_counter()
+    config_values = load_config_values()
     root_logger = logging.getLogger()
-    logger.info(configure_logger(root_logger, verbose))
+    log_level = config_values["LOG_LEVEL"] or "INFO"
```
Comment (hakbailey) [inline patronload/cli.py:17]: I don't think the `or "INFO"` is needed here since it defaults to `INFO` in config.

File: patronload/config.py (hunk 7 above)
Comment (ehanson8) [reply patronload/config.py:40]: I would prefer to have a reminder at the start of the process to set anything that is missing since this is a simple function. You're correct about `LOG_LEVEL` being optional so I'll shift that.

Comment (hakbailey) [reply patronload/config.py:40]: Sure, I'd still argue that adding an extra function for that is unneeded complexity. In alma-scripts we were configuring from both ENV and SSM which meant we had to build in extra checks. Now that we're getting everything from env, why not just use `os.environ[key]` and let the default `KeyError` bubble up?

Comment (ehanson8) [reply patronload/config.py:39]: Pushed a new commit to address this

**Review by hakbailey — APPROVED** [review]
Looks good!
